Bug: problem with projection! in jannovar-cli-033.jar

I get an error with the following variant in hg19 assembly:

##fileformat=VCFv4.1
##contig=<ID=1,length=249250621>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	individual
13	114435823	.	T	G	431.31	.	.	GT:AD:DP:GQ:PL	0/1:17,10:27:21:21,0,395

The error:

Options
JannovarAnnotateVCFOptions [escapeAnnField=true, pathInputVCF=ClinPrior.vcf, interval=, pathOutputVCF=outputJannovar.vcf, pathVCFDBSNP=null, prefixDBSNP=DBSNP_, pathFASTARef=null, pathVCFExac=null, prefixExac=EXAC_, pathVCFGnomadExomes=null, prefixGnomadExomes=GNOMAD_EXOMES_, pathVCFGnomadGenomes=null, prefixGnomadGenomes=GNOMAD_GENOMES_, pathVCFUK10K=null, prefixUK10K=UK10K_, pathThousandGenomes=null, prefixThousandGenomes=G1K_, pathClinVar=null, prefixClinVar=CLINVAR_, pathCosmic=null, prefixCosmic=COSMIC_, pathPedFile=null, annotateAsSingletonPedigree=false, useThresholdFilters=false, useAdvancedPedigreeFilters=false, threshFiltMinGtCovHet=8, threshFiltMinGtCovHomAlt=4, threshFiltMaxCov=10000, threshFiltMinGtGq=20, threshFiltMinGtAafHet=0.2, threshFiltMaxGtAafHet=0.8, threshFiltMinGtAafHomAlt=0.7, threshFiltMaxGtAafHomRef=0.3, threshFiltMaxExacHomAlt=20, threshFiltMaxThousandGenomesHomAlt=10, threshFiltMaxAlleleFrequencyAd=0.01, threshFiltMaxAlleleFrequencyAr=0.01, offTargetFilterEnabled=false, offTargetFilterUtrIsOffTarget=false, offTargetFilterIntronicSpliceIsOffTarget=false, oneParentGtFilteredFiltersAffected=false, inheritanceAnnoUseFilters=false, useParentGtIsFiltered=true, threshDeNovoParentAd2=1, bedAnnotationOptions=[], dbNsfpColContig=1, dbNsfpColPosition=2, prefixDbNsfp=DBNSFP_, pathDbNsfp=null, columnsDbNsfp=[], tsvAnnotationOptions=[], vcfAnnotationOptions=[]]
Deserializing transcripts...
INFO Deserializing JannovarData from C:/Users/aschluter.IDIBELL1573/Documents/R/win-library/3.6/ClinPrior/extdata/Jannovar/hg19_refseq.ser
INFO Deserialization took 7,37 sec.
Location	processed.sites	runtime.per.1M.sites	completed	total.runtime	remaining
Annotating VCF...
Will read full input file
Exception in thread "main" java.lang.Error: Bug: problem with projection!
	at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:117)
	at de.charite.compbio.jannovar.annotation.Annotation.toVCFAnnoString(Annotation.java:287)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.applyAnnotations(VariantContextAnnotator.java:400)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.dispatchAnnotateVariantContext(VariantContextAnnotator.java:337)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.annotateVariantContext(VariantContextAnnotator.java:296)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.Iterator.forEachRemaining(Unknown Source)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.forEachOrdered(Unknown Source)
	at de.charite.compbio.jannovar.cmd.annotate_vcf.AnnotateVCFCommand.run(AnnotateVCFCommand.java:411)
	at de.charite.compbio.jannovar.Jannovar.main(Jannovar.java:67)
Caused by: de.charite.compbio.jannovar.reference.ProjectionException: Invalid transcript sequence position 1302
	at de.charite.compbio.jannovar.reference.TranscriptProjectionDecorator.transcriptToGenomePos(TranscriptProjectionDecorator.java:169)
	at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:113)
	... 15 more
[1] 1

I get she same error on another VCF file (have to find the mutation which causes it). It is also refseq and hg19.

```
Exception in thread "main" java.lang.Error: Bug: problem with projection!
        at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:117)
        at de.charite.compbio.jannovar.annotation.Annotation.toVCFAnnoString(Annotation.java:287)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.applyAnnotations(VariantContextAnnotator.java:400)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.dispatchAnnotateVariantContext(VariantContextAnnotator.java:337)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.annotateVariantContext(VariantContextAnnotator.java:296)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:423)
        at de.charite.compbio.jannovar.cmd.annotate_vcf.AnnotateVCFCommand.run(AnnotateVCFCommand.java:411)
        at de.charite.compbio.jannovar.Jannovar.main(Jannovar.java:67)
Caused by: de.charite.compbio.jannovar.reference.ProjectionException: Invalid transcript sequence position 362
        at de.charite.compbio.jannovar.reference.TranscriptProjectionDecorator.transcriptToGenomePos(TranscriptProjectionDecorator.java:169)
        at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:113)

```

@holtgrewe any suggetsions?

Thanks for the fix. It works fine for Chr13:114435823 but I still have problems now for this variant (same error using updated hg19_refseq.ser):
1	148251986	rs199949374	T	G	2803.33	.	.	GT:AD:DP:GQ:PL	0/1:124,108:232:99:2816,0,3317

Exception in thread "main" java.lang.Error: Bug: problem with projection!
	at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:117)
	at de.charite.compbio.jannovar.annotation.Annotation.toVCFAnnoString(Annotation.java:287)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.applyAnnotations(VariantContextAnnotator.java:400)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.dispatchAnnotateVariantContext(VariantContextAnnotator.java:337)
	at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.annotateVariantContext(VariantContextAnnotator.java:296)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.Iterator.forEachRemaining(Unknown Source)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(Unknown Source)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.forEachOrdered(Unknown Source)
	at de.charite.compbio.jannovar.cmd.annotate_vcf.AnnotateVCFCommand.run(AnnotateVCFCommand.java:411)
	at de.charite.compbio.jannovar.Jannovar.main(Jannovar.java:67)
Caused by: de.charite.compbio.jannovar.reference.ProjectionException: Invalid transcript sequence position 15755
	at de.charite.compbio.jannovar.reference.TranscriptProjectionDecorator.transcriptToGenomePos(TranscriptProjectionDecorator.java:170)
	at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:113)
	... 15 more
[

Hi, 
I found that use annotate-pos does not return an error. However, the annotate-vcf returned an error. I also used the ser file you provided, but it didn't solve the problem.  Here's the log.
```
# md5sum 0.34_ser/hg38_refseq_curated.ser
de4777d32470517aa38819b8b513aef1  0.34_ser/hg38_refseq_curated.ser

# java -jar jannovar-cli-0.35.jar  annotate-vcf  -d 0.34_ser/hg38_refseq_curated.ser -i bug1.vcf -o tmp.vcf --show-all
Options
JannovarAnnotateVCFOptions [escapeAnnField=true, pathInputVCF=bug1.vcf, interval=, pathOutputVCF=tmp.vcf, pathVCFDBSNP=null, prefixDBSNP=DBSNP_, pathFASTARef=null, pathVCFExac=null, prefixExac=EXAC_, pathVCFGnomadExomes=null, prefixGnomadExomes=GNOMAD_EXOMES_, pathVCFGnomadGenomes=null, prefixGnomadGenomes=GNOMAD_GENOMES_, pathVCFUK10K=null, prefixUK10K=UK10K_, pathThousandGenomes=null, prefixThousandGenomes=G1K_, pathClinVar=null, prefixClinVar=CLINVAR_, pathCosmic=null, prefixCosmic=COSMIC_, pathPedFile=null, annotateAsSingletonPedigree=false, useThresholdFilters=false, useAdvancedPedigreeFilters=false, threshFiltMinGtCovHet=8, threshFiltMinGtCovHomAlt=4, threshFiltMaxCov=10000, threshFiltMinGtGq=20, threshFiltMinGtAafHet=0.2, threshFiltMaxGtAafHet=0.8, threshFiltMinGtAafHomAlt=0.7, threshFiltMaxGtAafHomRef=0.3, threshFiltMaxExacHomAlt=20, threshFiltMaxThousandGenomesHomAlt=10, threshFiltMaxAlleleFrequencyAd=0.01, threshFiltMaxAlleleFrequencyAr=0.01, offTargetFilterEnabled=false, offTargetFilterUtrIsOffTarget=false, offTargetFilterIntronicSpliceIsOffTarget=false, oneParentGtFilteredFiltersAffected=false, inheritanceAnnoUseFilters=false, useParentGtIsFiltered=true, threshDeNovoParentAd2=1, bedAnnotationOptions=[], dbNsfpColContig=1, dbNsfpColPosition=2, prefixDbNsfp=DBNSFP_, pathDbNsfp=null, columnsDbNsfp=[], tsvAnnotationOptions=[], vcfAnnotationOptions=[]]
Deserializing transcripts...
INFO Deserializing JannovarData from 0.34_ser/hg38_refseq_curated.ser
INFO Deserialization took 6.33 sec.
Location        processed.sites runtime.per.1M.sites    completed       total.runtime   remaining
Annotating VCF...
Will read full input file
Exception in thread "main" java.lang.Error: Bug: problem with projection!
        at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:117)
        at de.charite.compbio.jannovar.annotation.Annotation.toVCFAnnoString(Annotation.java:287)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.applyAnnotations(VariantContextAnnotator.java:400)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.dispatchAnnotateVariantContext(VariantContextAnnotator.java:337)
        at de.charite.compbio.jannovar.htsjdk.VariantContextAnnotator.annotateVariantContext(VariantContextAnnotator.java:296)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:502)
        at de.charite.compbio.jannovar.cmd.annotate_vcf.AnnotateVCFCommand.run(AnnotateVCFCommand.java:411)
        at de.charite.compbio.jannovar.Jannovar.main(Jannovar.java:67)
Caused by: de.charite.compbio.jannovar.reference.ProjectionException: Invalid transcript sequence position 3597
        at de.charite.compbio.jannovar.reference.TranscriptProjectionDecorator.transcriptToGenomePos(TranscriptProjectionDecorator.java:170)
        at de.charite.compbio.jannovar.annotation.VCFAnnotationData.setAnnoLoc(VCFAnnotationData.java:113)
        ... 15 more

  
```
[problematic_position.zip](https://github.com/charite/jannovar/files/6623646/problematic_position.zip)

I attached the vcf file that I used.  The position is different, but they locate at the same region and cause the same error.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,7 @@
 * Apply fix for (#498, PR #499 by @roland-ewald of @limbus-medtec).
   This fixes a problem with right-shifting deletions on amino acid sequences.
   See the tickes and merge request for details.
+* Fixing problem with GRCh38 RefSeq annotation / projection (#524).
 
 ### jannovar-cli
 

--- a/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/AnnotationLocation.java
+++ b/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/AnnotationLocation.java
@@ -146,4 +146,14 @@ public class AnnotationLocation {
 		return true;
 	}
 
+	@Override
+	public String toString() {
+		return "AnnotationLocation{" +
+			"transcript=" + transcript +
+			", rankType=" + rankType +
+			", rank=" + rank +
+			", totalRank=" + totalRank +
+			", txLocation=" + txLocation +
+			'}';
+	}
 }

--- a/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/VCFAnnotationData.java
+++ b/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/VCFAnnotationData.java
@@ -113,8 +113,7 @@ class VCFAnnotationData {
 			this.cdsPos = projector.projectGenomeToCDSPosition(projector.transcriptToGenomePos(txPos)).getPos();
 			this.cdsLength = transcript.cdsTranscriptLength();
 		} catch (ProjectionException e) {
-			// e.printStackTrace();
-			throw new Error("Bug: problem with projection!", e);
+			throw new Error("Bug: problem with projection! (" + annoLoc + ")", e);
 		}
 	}
 

--- a/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.java
+++ b/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptProjectionDecorator.java
@@ -166,7 +166,7 @@ public final class TranscriptProjectionDecorator {
 		if (targetSeqPos < 0) {
 			throw new ProjectionException("Invalid transcript sequence position " + targetSeqPos);
 		} else if (targetSeqPos
-			> Anchors.seqLength(transcript.getSeqAlignment().getRefAnchors()) + 1) {
+			> Anchors.seqLength(transcript.getSeqAlignment().getQryAnchors()) + 1) {
 			throw new ProjectionException("Invalid transcript sequence position " + targetSeqPos);
 		}
 

